specs/error_state/ErrorOutOfGasTloadTstore.md

```diff
@@ -0,0 +1,40 @@
+# ErrorOutOfGasTloadTstore state for both TLOAD and TSTORE OOG errors
+
+## Procedure
+
+Handle the corresponding out of gas errors for both `TLOAD` and `TSTORE` opcodes.
+
+### EVM behavior
+
+The out of gas error may occur for `constant gas`.
+
+#### TLOAD gas cost
+
+For this gadget, TLOAD gas cost in EIP-1153 is specified as the cost of hot SLOAD, currently `100`.
+
+#### TSTORE gas cost
+
+For this gadget, TSTORE gas cost in EIP-1153 is specified as the cost of SSTORE on an already SSTOREd slot, currently `100`.
+
+### Constraints
+
+1. For TLOAD, constrain `gas_left < gas_cost`.
+2. For TSTORE, constrain `gas_left < gas_cost`.
+3. Only for TSTORE, constrain `is_static == false`.
+4. Current call must fail.
+5. If it's a root call, it transits to `EndTx`.
+6. If it isn't a root call, it restores caller's context by reading to `rw_table`, then does step state transition to it.
+7. Constrain `rw_counter_end_of_reversion = rw_counter_end_of_step + reversible_counter`.
+
+### Lookups
+
+7 bus-mapping lookups for TLOAD and 8 for TSTORE:
```

Comment:
I think 7 is correct because it is in 92TLOAD_93TSTORE.md too? Or can you please correct me?

Reply:
Hm, 7 doesn't match the listed lookups below (5 call context lookups and 1 stack read). Note that for `TLOAD` we don't need `is_static`. But on the other hand, one stack operation seems to be missing (value read), so this is the reason for the difference in failure and success cases. But I agree, I think these two numbers should be the same.

Note, however, that the numbers in failure/success cases are different for `SLOAD/SSTORE` too. Is there a mistake?

One additional thing that I find a bit confusing is that the failure case is using  only `is_success` while the success case is using only `is_persistent`, shouldn't be used both in both cases?

Reply:
> Note that for TLOAD we don't need is_static. 

I agree. I saw SLOAD's OOG also having `is_static` in the [specs](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/09d127995a6b53c9ccd054478fea30452bdb8e2f/specs/error_state/ErrorOutOfGasSloadSstore.md?plain=1#L58). That needs to be fixed?

> But I agree, I think these two numbers should be the same. Note, however, that the numbers in failure/success cases are different for SLOAD/SSTORE too. Is there a mistake?

It would be very helpful if @ChihChengLiang or @ed255 could chime in regarding this.

> One additional thing that I find a bit confusing is that the failure case is using only is_success while the success case is using only is_persistent, shouldn't be used both in both cases?

Yeah the same name should be used. `is_persistent` seems to convey the idea better than `is_success`. I can update that.

Reply:
I will have another look, but just before you do any modifications - note that `is_persistent` and `is_success` are two different fields of `Call` struct which mean different things (is't not just the naming):
```
/// This call is persistent or call stack reverts at some point
pub is_persistent: bool,
/// This call ends successfully or not
pub is_success: bool,
```

Reply:
I think `is_success` and `is_persistent` are ok as they are (`is_success` needs to be false and is checked for each OOG error in `CommonErrorGadget`).

I think if you remove `is_static` lookup for from the error lookups for `SLOAD`, it should be fine (that would sum up to 5 `SLOAD` lookups in the error case).

`SLOAD/SSTORE` docs are ok in my opinion, except that for `SSLOAD` there is a redundant `is_static` in the error case.

Reply:
I agree that SLOAD OOG doesn't need to read `is_static`, and the same would happen for TLOAD OOG.  On the other hand, by looking at the implementation I see that `is_static` value is ignored in the SLOAD OOG case:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4996eb6c11fb4eb4bcde7eff55827ecf6cfaaf03/zkevm-circuits/src/evm_circuit/execution/error_oog_sload_sstore.rs#L70-L71
So the implementation is technically correct, even though we could skip the read of `is_static` in the SLOAD case.  I think this is a very minor detail; skipping the read means one less row in the RwTable, which is an optimization, but I think the effects are negligible.

>  But I agree, I think these two numbers should be the same.
> 
> Note, however, that the numbers in failure/success cases are different for `SLOAD/SSTORE` too. Is there a mistake?

Do you mean that `SLOAD.rwc_delta == OOG_SLOAD.rwc_delta` and `SSTORE.rwc_delta == OOG_SSTORE.rwc_delta`?

If that's the case, notice that in a success SLOAD, there's the value pushed to the stack, but in OOG, the call will be aborted, so no one is going to read the result of SLOAD, which means there's no need to push a result value to the stack.  So that's one RwTable lookup that appears in SLOAD but not in OOG_SLOAD.  There are other differences like in SLOAD we update the access list, but in OOG_SLOAD we don't.

Reply:
Thanks Edu! Yes, I think it's ok as it is (I realised later there's no additional push in the error case). Agree that the additional `is_static` call in TLOAD/SLOAD doesn't change things much, so I am approving the PR.